app-engine/java/ultimate/src/com/google/cloud/tools/intellij/appengine/java/ultimate/impl/AppEngineStandardUltimateWebIntegration.java

```diff
@@ -96,26 +107,38 @@ public VirtualFile suggestParentDirectoryForAppEngineWebXml(
     }
 
     for (WebRoot webRoot : webRoots) {
-      VirtualFile parent = webRoot.getFile();
-      if (parent != null) {
-        if (WEB_INF.equals(parent.getName())) {
-          return parent;
+      VirtualFile webRootDir = webRoot.getFile();
+      if (webRootDir != null) {
+        if (WEB_INF.equals(webRootDir.getName())) {
+          return webRootDir;
         }
 
-        VirtualFile child = parent.findChild(WEB_INF);
-        if (child != null) {
-          return child;
+        VirtualFile webInfDir = webRootDir.findChild(WEB_INF);
+        if (webInfDir != null) {
+          return webInfDir;
         }
       }
     }
 
     try {
-      VirtualFile webRootFile = webRoots.get(0).getFile();
-      if (webRootFile != null) {
-        return VfsUtil.createDirectoryIfMissing(webRootFile, WEB_INF);
+      Optional<VirtualFile> webRootDir =
```

Comment:
so here we're looking for a web root file that exists (physically) on the disk but doesn't contain WEB-INF, after initial loop on `webRoots`, correct?

Reply:
exactly. previously it was just grabbing the first one which may or may not have physically existed, which was not optimal

Reply:
could we add a small comment here to define a bit better what it means when a VirtualFile exists - meaning the root exists and we don't need to create it?